plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -41,20 +45,66 @@ abstractly specifiable. It's an implementation detail.
 
 -}
 
+{- Note [Integer types for costing]
+We care about the speed of our integer operations for costing, this has a significant effect on speed.
+But we also need to care about overflow: the cost counters overflowing is a potential attack!
+
+We have a few choices here for what to do with an overflow:
+- Don't (this is what 'Integer' does, it's unbounded)
+- Wrap (this is what 'Int'/'Int64' and friends do)
+- Throw an overflow error (this is what 'Data.SafeInt' does)
+- Saturate (i.e. return max/min bound, this is what 'Data.SatInt does)
+
+In our case
+- Not overflowing would be nice, but 'Integer' is significantly slower than the other types.
+- Wrapping is quite dangerous, as it could lead to us getting attacked by someone wrapping
+their cost around to something that looks below the budget.
+- Throwing would be okay, but we'd have to worry about exception catching.
+- Saturating is actually quite nice: we care about whether `a op b < budget`. So long as `budget < maxBound`,
+  then `a op b < budget` will have the same truth value *regardless* of whether the operation overflows and saturates,
+  since saturating implies `a op b >= maxBound > budget`. Plus, it means we don't need to deal with
+  exceptions.
+
+So we use 'Data.SatInt', a variant of 'Data.SafeInt' that does saturating arithmetic.
+
+'SatInt' is quite fast, but not quite as fast as using 'Int64' directly (I don't know why that would be, apart from maybe
+just the overflow checks), but the wrapping behaviour of 'Int64' is unacceptable..
+
+One other wrinkle is that 'SatInt' is backed by an 'Int' (i.e. a machine integer with platform-dependent
```

Comment:
I don't understand the issue here. Do we want to fix our 'integer-type' to be `Int64` but SafeInt is written using `Int`? If this is the issue, wouldn't something like replacing `Int ==> Int64` text in the SatInt resolve it?

Reply:
As the next sentence says, the primops we need aren't available for `Int64` until later GHCs :shrug: